jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
+    }
+    if sortkey is None:  # mimic default ps commnad sorting behavior
+        sortkey = "pid"
+    # return pids from process objects
+    return [p["pid"] for p in get_processes_info(sort=ps_to_psutil_map[sortkey], filterstr=filterstr, desc=desc)]
 
 
 def get_filtered_pids(filterstr, excludes=None):
-    """Get pids filtered by filterstr and execludes
-
-    Arguments:
-        filterstr {str} -- filter string.
+    """Get pids filtered by filterstr and excludes, matching against the full command line used to start the process.
 
-    Keyword Arguments:
-        excludes {list(str)} -- execlude list (default: {None})
+    Args:
+        filterstr (str): the String to filter based on.
+        excludes (list[str]): exclude list. Defaults to None.
 
     Returns:
-        [list(int)] -- pids
-    """
-    excludes = excludes or []
-    cmd = "ps ax | grep '%s'" % filterstr
-    rc, out, err = j.core.executors.run_local(cmd)
-    # print out
-    found = []
-
-    def checkexclude(c, excludes):
-        for item in excludes:
-            c = c.lower()
-            if c.find(item.lower()) != -1:
-                return True
-        return False
-
-    for line in out.split("\n"):
-        if line.find("grep") != -1 or line.strip() == "":
-            continue
-        if line.strip() != "":
-            if line.find(filterstr) != -1:
-                line = line.strip()
-                if not checkexclude(line, excludes):
-                    # print "found pidline:%s"%line
-                    found.append(int(line.split(" ")[0]))
-    return found
-
+        list of int: list of the processes IDs
+    """
+    pids = []
+    try:
+        for proc in psutil.process_iter(["name", "cmdline"]):
+            cmd_line = " ".join(proc.info["cmdline"])
+            if proc.info["cmdline"] and filterstr in cmd_line:
+                j.logger.debug(f"found filter string: {filterstr} in command line: {cmd_line}")
+                if excludes:
+                    for exclude in excludes:
+                        if exclude in cmd_line:
+                            j.logger.debug(
+                                f"but excluded because it contain exclude string: {exclude} in command line: {cmd_line}"
+                            )
+                            break
+                    else:  # intended `for/else` meaning for loop finished normally with no break
+                        pids.append(proc.pid)  # may yield proc instead
+                    continue
+                else:
+                    pids.append(proc.pid)  # may yield proc instead
+        j.logger.debug(
+            f"founded pids are: {pids}"
+            if pids
+            else f"filter string {filterstr} not found in any processes. root needed?"
+        )
+        return pids
+    except (psutil.AccessDenied, psutil.NoSuchProcess) as e:
+        j.logger.exception("exception occurred while iterating over the system processes", exception=e)
+        pass
```

Comment:
Should return an empty list instead?

```suggestion
        return []
```

Reply:
no. actually this is not a return statement, this try/ except block should be inside the for loop. and it should continue to the next iteration when an exception raised. Thanks, fixed.